File: 크레인인형뽑기게임/java/20221104/src/Solution.java
import java.util.*;

class Solution {
    public void test() {
        int[][] board = new int[][]{{0,0,0,0,0},{0,0,1,0,3},{0,2,5,0,1},{4,2,4,4,2},{3,5,1,3,1}};
        int[] moves = new int[]{1,5,3,5,1,2,1,4};

        System.out.println("process(board, moves): 4 = " + process(board, moves));

        board = new int[][]{{0,0,0,0,0},{0,0,1,0,3},{0,2,5,0,1},{4,2,4,4,2},{3,5,1,3,1}};
        moves = new int[]{1,4,1,4};

        System.out.println("process(board, moves): 4 = " + process(board, moves));
    }

    public int solution(int[][] board, int[] moves) {
        // test();
        int answer = process(board, moves);
        return answer;
    }

    public int process(int[][] board, int[] moves) {
        List<Integer> bag = new ArrayList<>();
        int count = 0;

        for (int i : moves) {
            for (int j = 0; j < board.length; j += 1) {
                int value = board[j][i - 1];

                if (value > 0) {
                    board[j][i - 1] = 0;

                    if (bag.size() > 0 && bag.get(bag.size() - 1) == value) {
                        bag.remove(bag.size() - 1);

                        count += 2;

                        break;
                    }

                    bag.add(value);

                    break;
                }
            }
        }

        return count;
    }
}

File: 크레인인형뽑기게임/javascript/20221104/src/solution.js
function test() {
  const board = [[0, 0, 0, 0, 0], [0, 0, 1, 0, 3], [0, 2, 5, 0, 1], [4, 2, 4, 4, 2], [3, 5, 1, 3, 1]];
  const moves = [1, 4];

  console.log(`process(board,moves): 2 = ${process(board, moves)}`);
}

function solution(board, moves) {
  // test();
  const answer = process(board, moves);
  return answer;
}

function process(board, moves) {
  let count = 0;
  let bag = [];

  for (i of moves) {
    for (let j = 0; j < board.length; j += 1) {
      const value = board[j][i - 1];

      if (value > 0) {
        board[j][i - 1] = 0;

        if (bag[bag.length - 1] === value) {
          bag.pop();
          count += 2;
          break;
        }

        bag = [...bag, value];

        break;
      }
    }
  }

  return count;
}
